lightning/src/ln/peers/handler.rs

```diff
@@ -60,6 +60,9 @@ pub(super) trait ITransport {
 	/// Returns true if the connection is established and encrypted messages can be sent.
 	fn is_connected(&self) -> bool;
 
+	/// Returns all Messages that have been received and can be parsed by the Transport
+	fn drain_messages<L: Deref>(&mut self, logger: L) -> Result<Vec<Message>, PeerHandleError> where L::Target: Logger;
```

Comment:
Done: `review: Small nits, documentation, and fixups`

Reply:
Think this got lost in the noise.

This was the design chosen in #494 and I chose to limit the number of behavior changes in the initial refactoring to ease the review (although I'm not sure how much that really saved in the end). I've done my best to separate out behavior changes to a separate patch so they can be easily dropped if necessary.

This could just as easily be a drain iterator or an iterator followed by clear() or drain(). I don't really like the pattern of iterators that also remove, so this was the cleanest analog. I don't think there is any additional memory pressure here since the bytes exist in the Decryptor anyway and they are effectively transferred to this layer.

But, it seems reasonable to change this up in the future if there is a valid reason. An iterator/drain pattern would mesh well with my thoughts in #698.

Reply:
Am I reading it wrong in `transport.rs`? It looks like it processes bytes, creating messages and copying them onto the heap (in the vec), possible with reallocation (ie copies) to grow the vec. An iterator/drain iterator would just deserialize directly onto the stack and return.

Reply:
I think it is probably worth looking at the code after 2b8d24d and recommending your changes on top of it. At that point, the code smells around iterators that can return errors and the feature to process the undecrypted message directly from the input slice has been implemented.

The original code from 494 had a bug in that it just asserted on all errors. The initial patches fixed this up to follow a more standard iterator-that-can-error pattern, but as you will see looking at it, it is definitely a code smell and hard to maintain.

Future patches culminating in 2b8d24d made this much easier to test and reason about. If you aren't happy with the performance characteristics that is probably a good place to talk about the tradeoffs so that the right design/fix can go in.